reorganize uploader docs (#615)
Laravel-Backpack/docs: 7.x-dev/crud-uploaders.md

<a name="available-uploaders"></a>
## Available Uploaders

We've already created Uploaders for the most common scenarios:
- CRUD comes with `SingleFile`, `MultipleFiles`, `SingleBas64Image`
- PRO comes with `DropzoneUploader`, `EasyMDEUploader`, `SummernoteUploader`
- if you want to use spatie/medialibrary you can just install [medialibrary-uploaders](https://github.com/Laravel-Backpack/medialibrary-uploaders) to get `MediaAjaxUploader`, `MediaMultipleFiles`, `MediaSingleBase64Image`, `MediaSingleFile`

<a name="how-to-create-uploaders"></a>
## How to Create Uploaders

Do you want to create your own Uploader class, for your custom field? Here's how you can do that, and how Uploader classes work behind the scenes.

First thing you need to decide if you need "Ajax" or "Non-Ajax" upload. The big difference is that the "non-ajax" uploaders process the file upload when you submit your form, while the ajax upload process the file using a javascript ajax request, so it can be uploaded before you submit the form. 

> 

Author: ```suggestion

<a name="how-to-create-a-custom-non-ajax-uploader"></a>
## How to Create a Custom Non-Ajax Uploader

```

First let's see how to create a non-ajax uploader, for that we will create a `CustomUploader` class that extends the abstract class `Uploader`. 

```php
namespace App\Uploaders\CustomUploader;

use Backpack\CRUD\app\Library\Uploaders\Uploader;

class CustomUploader extends Uploader
{
    // the function we need to implement
    public function uploadFiles(Model $entry, $values)
    {
        // $entry is the model instance we are working with
        // $values is the sent files from request.

        // do your upload logic here

        return $valueToBeStoredInTheDatabaseEntry;
    }

    // in case you want to use your uploader inside repeatable fields
    protected function uploadRepeatableFiles($values, $previousValues, $entry = null)
    {

>     }

Author: This needs a little bit more info inside it. Ok so it's a method for repeatables, but what should I do inside it, to make an upload work inside repeatable subfields? I have no idea if you don't tell me.

}
```

You can now use this uploader in your field definition:
    
```php
CRUD::field('avatar')->type('upload')->withFiles([
    'uploader' => \App\Uploaders\CustomUploader::class,
]);
```

> But most likely, you have a `custom_upload` field that you'd like to use this uploader without having to specify it every time. You can do that by adding it to the `UploadersRepository` in your Service Provider `boot()` method:

Author: ```suggestion
If you custom uploader was created to work for a custom field (say it's called `custom_upload`), you can tell Backpack to always use this uploader for that field type - that way you don't have to specify it every time you use the field. You can do that in your Service Provider `boot()` method, by adding it to the `UploadersRepository`:
```

```php
// in your App\Providers\AppServiceProvider.php

protected function boot()
{
    app('UploadersRepository')->addUploaderClasses(['custom_upload' => \App\Uploaders\CustomUploader::class], 'withFiles');
}
```

> You can now use `CRUD::field('avatar')->type('custom_upload')->withFiles();` and it will use your custom uploader. What happen behind the scenes is that Backpack will register your uploader to be ran in 3 different model events: `saving`, `retrieved` and `deleting`.

Author: ```suggestion
You can now use `CRUD::field('avatar')->type('custom_upload')->withFiles();` and it will use your custom uploader. What happens behind the scenes is that Backpack will register your uploader to run on 3 different model events: `saving`, `retrieved` and `deleting`.
```

Author: Woow that's huge info. That should be up at the top too!

> The `Uploader` class has 3 "entry points" for the mentioned events: **`storeUploadedFiles()`**, **`retrieveUploadedFiles()`** and **`deleteUploadedFiles()`**. You can overwrite these methods in your custom uploader to add your custom logic but for most uploaders you will not need to overwrite them as they are "setup" methods for the action that will be performed, and after setup they call the relevant methods that each uploader will implement, like ```uploadFiles()``` or ```uploadRepeatableFiles()```.

Author: > for most uploaders you will not need to overwrite them as they are "setup" methods for the action that will be performed, and after setup they call the relevant methods that each uploader will implement, like ```uploadFiles()``` or ```uploadRepeatableFiles()```.

I don't understand this. Can you please explain it to me @pxpm ?

Reviewer: It's what I've explained in some other place about the "strategy methods". Instead of craming all the code to store an upload inside the "event method" (storeUploadedFiles). That method receives the event and delegate what will happen to the relevant functions. Eg: it checks if it's a repeatable, if it is, it call the "upload repeable files", otherwise it calls the "upload files". 

That's why that method usually don't need to be overriden, because there is no specific "upload functionality" on it, it's just delegation of tasks.

Author: ```suggestion
The `Uploader` class has 3 "entry points" for the mentioned events: **`storeUploadedFiles()`**, **`retrieveUploadedFiles()`** and **`deleteUploadedFiles()`**. You can override these methods in your custom uploader, but typically you will not need to do that. The methods already delegate what will happen to the relevant methods (eg. if it's not a repeatable, call ```uploadFiles()```, othewise call ```uploadRepeatableFiles()```).
```

> The base uploader class has most of the functionality implemented and use **"strategy methods"** to configure the underlying behavior. 

Author: ```suggestion
Notice this custom class you're creating is extending `Backpack\CRUD\app\Library\Uploaders\Uploader`. That base uploader class has most of the functionality implemented and uses **"strategy methods"** to configure the underlying behavior. 
```

**`hasDeletedFiles`** - a method that returns a boolean to determine if the files were deleted from the field.

**`getUploadedFilesFromRequest`** - this is the method that will be called to get the values sent in the request. Some uploaders require you get the `->files()` others the `->input()`. By default it returns the `->files()`.

This is the implementation of those methods in `SingleFile` uploader:
```php
protected function shouldKeepPreviousValueUnchanged(Model $entry, $entryValue): bool
{
    // if a string is sent as the value, it means the file was not changed so we should keep
    // previous value unchanged
    return is_string($entryValue);
}

protected function hasDeletedFiles($entryValue): bool
{
    // if the value is null, it means the file was deleted from the field
    return $entryValue === null;
}

protected function shouldUploadFiles($value): bool
{
    // when the value is an instance of UploadedFile, it means the file was uploaded and we should upload it
    return is_a($value, 'Illuminate\Http\UploadedFile', true);
}

> ```
> 
> For the ajax uploaders, the process is similar, but the `CustomUploader` class should extend `BackpackAjaxUploader` **(requires backpack/pro)** instead of `Uploader`.

Author: ```suggestion

<a name="how-to-create-a-custom-ajax-uploader"></a>
## How to Create a Custom Ajax Uploader

For the ajax uploaders, the process is similar, but your custom uploader class should extend `BackpackAjaxUploader` instead of `Uploader` (**note that this requires backpack/pro**).
```

```php

namespace App\Uploaders\CustomUploader;

use Backpack\Pro\Uploaders\BackpackAjaxUploader;

class CustomUploader extends BackpackAjaxUploader
{
    // this is called on `saving` event of the main entry, at this point you already performed the upload 
    // of the files in the ajax endpoint. By default they are in a temp folder, so here is the place 
    // where you should move them to the final disk and path and setup what will be saved in the database.
    public function uploadFiles(Model $entry, $values)
    {
        return $valueToBeStoredInTheDatabaseEntry;
    }

    // in case you want to use your uploader inside repeatable fields
    protected function uploadRepeatableFiles($values, $previousValues, $entry = null)
    {

>     }

Author: Same here, no idea what to do for repeatables, if you don't tell me.

}
```

The process to register the uploader in the `UploadersRepositoy` is the same as the non-ajax uploader. `app('UploadersRepository')->addUploaderClasses(['custom_upload' => \App\Uploaders\CustomUploader::class], 'withFiles');` in the boot method of your provider.

In addition to the field configuration, ajax uploaders require that you use the `AjaxUploadOperation` trait in your controller. The operation is responsible to register the ajax route where your files will be sent and the upload process will be handled and the delete route from where you can delete **temporary files**.

Similar to model events, there are two "setup" methods for those endpoints: **`processAjaxEndpointUploads()`** and **`deleteAjaxEndpointUpload()`**. You can overwrite them to add your custom logic but most of the time you will not need to do that and just implement the `uploadFiles()` and `uploadRepeatableFiles()` methods.

The ajax uploader also has the same "strategy methods" as the non-ajax uploader (see above), but adds a few more:

> **`ajaxEndpointSuccessResponse($files = null)`** - this should return a `JsonResponse` with the needed information when the upload is successful. By default it returns a json response with the file path.
> 
> **`ajaxEndpointErrorResponse($message)`** - use this method to change the endpoint response in case the upload failed. Similar to the success it should return a `JsonResponse`.
> 
> **`getAjaxEndpointDisk()`** - by default a `temporaryDisk` is used to store the files before they are moved to the final disk. (when uploadFiles() is called). You can overwrite this method to change the disk used.
> 
> **`getAjaxEndpointPath()`** - by default the path is `/temp` but you can overwrite this method to change the path used.
> 
> **`getDefaultAjaxEndpointValidation()`** - Should return the default validation rules (in the format of `BackpackCustomRule`) for the ajax endpoint. By default it returns a `ValidGenericAjaxEndpoint` rule.

Author: ```suggestion
- **`ajaxEndpointSuccessResponse($files = null)`** - This should return a `JsonResponse` with the needed information when the upload is successful. By default it returns a json response with the file path.
- **`ajaxEndpointErrorResponse($message)`** - Use this method to change the endpoint response in case the upload failed. Similar to the success it should return a `JsonResponse`.
- **`getAjaxEndpointDisk()`** - By default a `temporaryDisk` is used to store the files before they are moved to the final disk (when uploadFiles() is called). You can overwrite this method to change the disk used.
- **`getAjaxEndpointPath()`** - By default the path is `/temp` but you can override this method to change the path used.
- **`getDefaultAjaxEndpointValidation()`** - Should return the default validation rules (in the format of `BackpackCustomRule`) for the ajax endpoint. By default it returns a `ValidGenericAjaxEndpoint` rule.
```

<a name="available-uploaders"></a>
## Available Uploaders

We've already created Uploaders for the most common scenarios:
- CRUD comes with `SingleFile`, `MultipleFiles`, `SingleBas64Image`
- PRO comes with `DropzoneUploader`, `EasyMDEUploader`, `SummernoteUploader`
- if you want to use spatie/medialibrary you can just install [medialibrary-uploaders](https://github.com/Laravel-Backpack/medialibrary-uploaders) to get `MediaAjaxUploader`, `MediaMultipleFiles`, `MediaSingleBase64Image`, `MediaSingleFile`

<a name="how-to-create-uploaders"></a>
## How to Create Uploaders

Do you want to create your own Uploader class, for your custom field? Here's how you can do that, and how Uploader classes work behind the scenes.

First thing you need to decide if you are creating a _non-ajax_ or _ajax_ uploader:
- _non-ajax_ uploaders process the file upload when you submit your form; 
- _ajax_ uploaders process the file upload before the form is submitted, by submitting an AJAX request using Javascript; 

> <a name="how-to-create-a-custom-non-ajax-uploader"></a>
> ## How to Create a Custom Non-Ajax Uploader

Author: ```suggestion
<a name="how-to-create-a-custom-non-ajax-uploader"></a>
### How to Create a Custom Non-Ajax Uploader
```

**`hasDeletedFiles`** - a method that returns a boolean to determine if the files were deleted from the field.

**`getUploadedFilesFromRequest`** - this is the method that will be called to get the values sent in the request. Some uploaders require you get the `->files()` others the `->input()`. By default it returns the `->files()`.

This is the implementation of those methods in `SingleFile` uploader:
```php
protected function shouldKeepPreviousValueUnchanged(Model $entry, $entryValue): bool
{
    // if a string is sent as the value, it means the file was not changed so we should keep
    // previous value unchanged
    return is_string($entryValue);
}

protected function hasDeletedFiles($entryValue): bool
{
    // if the value is null, it means the file was deleted from the field
    return $entryValue === null;
}

protected function shouldUploadFiles($value): bool
{
    // when the value is an instance of UploadedFile, it means the file was uploaded and we should upload it
    return is_a($value, 'Illuminate\Http\UploadedFile', true);
}

> <a name="how-to-create-a-custom-ajax-uploader"></a>
> ## How to Create a Custom Ajax Uploader

Author: ```suggestion
<a name="how-to-create-a-custom-ajax-uploader"></a>
### How to Create a Custom Ajax Uploader
```
